sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/storage/GcsResourceIdTest.java

```diff
@@ -0,0 +1,121 @@
+
+  @Test
+  public void testResolve() throws Exception {
+    // Tests for common gcs paths.
+    assertEquals(
+        toResourceIdentifier("gs://bucket/tmp/aa"),
+        toResourceIdentifier("gs://bucket/tmp/")
+            .resolve("aa", StandardResolveOptions.RESOLVE_FILE));
+    assertEquals(
+        toResourceIdentifier("gs://bucket/tmp/aa/bb/cc/"),
+        toResourceIdentifier("gs://bucket/tmp/")
+            .resolve("aa", StandardResolveOptions.RESOLVE_DIRECTORY)
+            .resolve("bb", StandardResolveOptions.RESOLVE_DIRECTORY)
+            .resolve("cc", StandardResolveOptions.RESOLVE_DIRECTORY));
+
+    // Tests absolute path.
+    assertEquals(
+        toResourceIdentifier("gs://bucket/tmp/aa"),
+        toResourceIdentifier("gs://bucket/tmp/bb/")
+            .resolve("gs://bucket/tmp/aa", StandardResolveOptions.RESOLVE_FILE));
+
+    // Tests path with unicode
+    assertEquals(
+        toResourceIdentifier("gs://bucket/输出 目录/输出 文件01.txt"),
+        toResourceIdentifier("gs://bucket/输出 目录/")
+            .resolve("输出 文件01.txt", StandardResolveOptions.RESOLVE_FILE));
+  }
+
+  @Test
+  public void testResolveHandleBadInputs() throws Exception {
+    assertEquals(
+        toResourceIdentifier("gs://my_bucket/tmp/"),
+        toResourceIdentifier("gs://my_bucket/")
+            .resolve("tmp/", StandardResolveOptions.RESOLVE_DIRECTORY));
+  }
+
+  @Test
+  public void testResolveInvalidInputs() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage(
+        "The resolved file: [tmp/] should not end with '/'.");
+    toResourceIdentifier("gs://my_bucket/").resolve("tmp/", StandardResolveOptions.RESOLVE_FILE);
+  }
+
+  @Test
+  public void testResolveInvalidNotDirectory() throws Exception {
+    ResourceId tmpDir = toResourceIdentifier("gs://my_bucket/")
+        .resolve("tmp dir", StandardResolveOptions.RESOLVE_FILE);
+
+    thrown.expect(IllegalStateException.class);
+    thrown.expectMessage("Expected the gcsPath is a directory, but had [gs://my_bucket/tmp dir].");
+    tmpDir.resolve("aa", StandardResolveOptions.RESOLVE_FILE);
+  }
+
+  @Test
+  public void testGetCurrentDirectory() throws Exception {
+    // Tests for local files without the scheme.
+    assertEquals(
+        toResourceIdentifier("gs://my_bucket/tmp dir/"),
+        toResourceIdentifier("gs://my_bucket/tmp dir/").getCurrentDirectory());
+
+    // Tests path with unicode
+    assertEquals(
+        toResourceIdentifier("gs://my_bucket/输出 目录/"),
+        toResourceIdentifier("gs://my_bucket/输出 目录/文件01.txt").getCurrentDirectory());
+  }
+
+  @Test
+  public void testEquals() throws Exception {
+    assertEquals(
+        toResourceIdentifier("gs://my_bucket/tmp/"),
+        toResourceIdentifier("gs://my_bucket/tmp/"));
+
+    assertNotEquals(
+        toResourceIdentifier("gs://my_bucket/tmp"),
+        toResourceIdentifier("gs://my_bucket/tmp/"));
+  }
```

Comment:
I think isDirectory() is the implementation details.

I added more tests to make sure public methods work correctly.